### DISCUSSION TITLE
Recover corrupted file

### DISCUSSION THREAD

**Discussion by RaphaelNeubert** [xournalpp/xournalpp#3056] [Q&A]
Hello,
I've screwed up my timing while running some rsync command and accidentely deleted a .xopp file with information i need.
Unfortunataly this was the only file on the system that wasn't backedup and since rsync deleted it i couldn't find it in any recycling bin. In order to restore it i ran Disk Drill and it found a total of 90 .xopp files. They are all numbered so basically <filename>1.xopp, <filename>2.xopp, .... Half of the 90 .xopp files are autosaves. When i try to open them i get Errors. For example "XML Parser error: Error on line 8 char 3813: Odd character "k" expected a">" characcter to end the empty-element tag"YSH"" I opened the file with vim but its just binary.

What gives me hope is that when i open xournalpp, go to open and select the file in the file manager, i still see a preview of the file. I don't really need the file back, i just need to take a look at the information inside of it. Sadly i couln't maximize the preview or do something like that. 

I appreciate any help.

**Comment by personalizedrefrigerator** [comment]
> I opened the file with vim but its just binary.

`.xopp` files are GZipped XML files! Try renaming the file to something with an `.xml.gz` extension, then unzipping it.

On Linux/Unix:
```sh
$ mv foo.xopp foo.xml.gz
$ gunzip foo.xml.gz
$ vim foo.xml
... Fix errors ...
$ gzip foo.xml
$ mv foo.xml.gz foo.xournalpp
```

**Comment by personalizedrefrigerator** [comment]
Alternatively, from within `Vim`, 
 * Open the `.xopp` file.
 * Run `:%! gunzip`
    * You should now have an XML file.
    * There will probably be a `gzip: unexpected end of file` line at the end of the file. **Remove this.**
 * Run `:set filetype=xml` :)
 * Fix any errors in the XML.
 * Run `:%! gzip` and write!
 * Open the `.xournalpp` file.

**Reply by personalizedrefrigerator** [reply]
If there are too many errors to easily fix by hand, you can enlarge the preview (but, when I did this, it wasn't very readable)!
 * Follow the above steps to get an XML representation of the file.
 * You should see a b64-encoded string in a `<preview></preview>` tag. For example,
![Example preview tag](https://user-images.githubusercontent.com/46334387/115505639-9f50cc80-a22e-11eb-96e2-db7e1bec04a4.png)
 * Make a new `.xopp` file with just an image in it. Turn it into XML using the steps above.
     * It should look something like this:
 
```xml

  <?xml version="1.0" standalone="no"?>    
  <xournal creator="Xournal++ 1.1.0+dev" fileversion="4">    
  <title>Xournal++ document - see https://github.com/xournalpp/xournalpp</title>    
  <preview>iVBORw0KGgoAAAANSUhEUgAAAFoAAACACAIAAAAEb+kcAAAABmJLR0QA/wD/AP+gvaeTAAADqUlEQVR4nO2dzZKbMBCEp4VzT+WWUy55gbz/qy3qHPgxMgiNbGwsb3elVGtqaIz0aWQYVQV///wzaVI4+wu8l8buAMqtN6wq+GBP+D03v0D36+fvbF99P4kO0ZHXTAeKrTMMSMZqx63K017ieRnOJ5lvGdEj0Mw4BDva4Q8ykIBh7Txr99Jz6wp+3PMydaeR2200WhcjeqsXYgj9D+b99y99DcPYzaXgsd89nptfYKYj2yJYNCYwRjOG9W8WgoYILAYKFo0w5Pz3L30N8wbT77n5BXwry819Gyx26C/2dVm2IYZlV3icy5d+7cpSpsNgXPdIiOOUG04fWkTiNpI0uAenSToIMvR2+bLua9kyxEUmNY9z+dJ1wS+jg2axM2LdO9c7Tz5EhoiD6HAHP0rHJXNrKzFY7BDdj3yht3UieXvVTRZ/2qO5uK3ydAQ/OlnG7ihwNXVHVdrznFLr+V6ptGokPafUehaDrz/AHZ7r4A+jw5Y/wIue6+APo0O5Q7lDdDiDRYdWFq0szmDljnfPHYtR8gVXeN4x4FWex9MBYEnIIXTMnncMeJWn+wHfrZFYlOLu8DxUm55KpUqlL02louPz6MBea6DRxuJTsaWBmF4PHlX3fZsaLYzDvXV9jHEeqv3WQBvewJObNdrFSOYvfW3NEzzriTVaDp0IGli9d4o2rO45/8KlX16jrcsdd6jR3JFtSQONxB3/jGhrZXFU4QzoO4t3IUKQQDt0jJuhSkm4MwYwBOus3AZYmD4iIGx6znqrlUV7wxJpm20ibaRMPmqyJBIdoiMvbbPVQpuXcodyR16iQ3Tk1QYd7mDRcaieQMc0Smd4VtCx6Tk/0WKnvXlTsh8M2PS65kBP+Dznd1R7YbMnV56aLInaSKUv2wwlOhJ9GB0VqVQLbVmiQ3Tk9WF0aGU5VB9Gh3LHodKr4+Tg8UVJ8vp4tqObGyh6ejaZPO6pyZJIqVSpNC/RITryaoMOd7DoOFRt0KHccY5Eh+jIS49wyUHRkUg12uTg+TXa/bDJE6rRniAttFpo8zqRjkdHUnQ8XcodoiMv0SE68mqDDq0s50gP+MlB1WgTabIkaiOVajPUOWqDDi2056gNOvQj/RyJDtGRl+gQHXm1QYc7WHQcqqfUaH3BFSM5xz/b01WjvU4AV+1zvt5eWGU91db11Ic9VaMtqY1UqoX2HLVBhxbac9QGHcod50h0iI68VIVLDoqORKIjOfita7Sr/1pZkyWVFtrk438svja7RaWRnQAAAABJRU5ErkJggg==</preview>
  <page width="595.27559100" height="841.88976400">    
  <background type="solid" color="#241f31ff" style="dotted"/>    
  <layer>    
  <image left="1.26784327" top="11.12237871" right="287.71546456" bottom="297.57000000">iVBORw0KGgoAAAANSUhEUgAAABYAAAAWCAYAAADEtGw7AAAABmJLR0QA/wD/AP+gvaeTAAAARklEQVQ4jWNgGAVQwIhLQllN8z8xBty9dR2rGSzkaCbGUqIMJtb1RAFiDMOnholqLhk1eNTgUYOJB3jLCkrKCJq5eBTAAQCvMQ3oIOcLXAAAAABJRU5ErkJggg==</image>
  </layer>    
  </page>    
  </xournal>
```

  * Replace the contents of the `<image ...> </image>` tag with the contents of the preview from the damaged file.
  * Run `:%! gzip` (make sure you've removed the `gzip: unexpected end of file` first!) and write.
  * Open it with Xournal++.

When I did this, the output was not very readable:
![image](https://user-images.githubusercontent.com/46334387/115506634-f1462200-a22f-11eb-9c93-6afbaa2fbc50.png)

**Comment by RaphaelNeubert** [comment]
Thanks for the help.

I've tried the stuff above and was able to see and edit the xml stuff. Unfortunately the data seems to be very very corrupted. I couldn't fix it. The only thing that wasn't corrupted in some files was the preview. Of course its not readable but there was a lot of graph theorie stuff on there so i can  at least see the different nodes which will help me a lot to resolve the tasks.
